lib/src/worker/worker.dart

```diff
@@ -0,0 +1,117 @@
+// ignore_for_file: public_member_api_docs
+
+import 'dart:async';
+import 'dart:convert' show jsonEncode;
+import 'dart:js_interop';
+import 'dart:js_interop_unsafe';
+// ignore: avoid_web_libraries_in_flutter
+import 'dart:js_util';
+
+import 'package:meta/meta.dart';
+import 'package:web/web.dart' as web;
+
+// Masked type: ServiceWorkerGlobalScope
+@JS('self')
+external JSAny get globalScopeSelf;
+
+@JS('self.importScript')
+// ignore: unused_element
+external JSAny _importScript(String path);
+
+void jsSendMessage(dynamic m) {
+  globalContext.callMethod('postMessage'.toJS, (m as Object).jsify());
+}
+
+Stream<T> callbackToStream<J, T>(
+  dynamic object,
+  String name,
+  T Function(J jsValue) unwrapValue,
+) {
+  final controller = StreamController<T>.broadcast(sync: true);
+  setProperty(
+    object as Object,
+    name,
+    allowInterop((J event) {
+      controller.add(unwrapValue(event));
+    }),
+  );
+  return controller.stream;
+}
+
+@internal
+class Worker {
+  Worker() {
+    _outputController = StreamController();
+    callbackToStream(globalScopeSelf, 'onmessage', (web.MessageEvent e) {
+      _outputController.add(getProperty<dynamic>(e as Object, 'data'));
+    });
+  }
+  late StreamController<dynamic> _outputController;
+
+  Stream<dynamic> onReceive() => _outputController.stream;
+
+  void sendMessage(dynamic message) {
+    jsSendMessage(message);
+  }
+}
+
+@internal
+class WorkerController {
+  web.Worker? _worker;
+  StreamController<dynamic>? _outputController;
+
+  /// Spawn a new web Worker with the given JS source (not used now).
+  static Future<WorkerController> spawn(String path) async {
+    final controller = WorkerController()
+      .._outputController = StreamController()
+      .._worker = web.Worker(path.endsWith('.dart') ? '$path.js' : path);
+
+    controller._worker?.onmessage = ((web.MessageEvent event) {
+      controller._outputController?.add(event.data.dartify());
+    }).toJS;
+
+    return controller;
+  }
+
+  /// Set the worker created in WASM.
+  /// This is used to get events sent from the native audio thread.
+  void setWasmWorker(web.Worker wasmWorker) {
```

Comment:
I feel like I need a diagram somewhere in the docs (maybe in `WEB_NOTES.md`?). What are the different workers (WASM worker, this worker), how do they communicate, with whom, etc. Either an ASCII graph or even just a piece of paper that you take a photo of.

Reply:
Sure, I'll try!

Reply:
I drew a sketch [here](https://excalidraw.com/#room=09413bf7417b2abd5b22,jJNb4nLgpMKf4vqOXZvmcA). It would be nice to have your opinion before exporting it as an image to insert it in WEB_NOTES. Feel free to modify it!

Reply:
This is really helpful! I don't think I have anything to change. It would be great to have things linked (like being able to actually see the EM_ASM() code, or linking to explanatory articles about WASM Modules), but I have no idea how we could do this without making the docs unnecessarily complicated. I think the current version is excellent.